[WIP] Homie v3.0.0

bodiroga opened the pull request:
Hi @jalmeroth!

This PR is a WIP to update the library to the latest Homie version: 3.0.0. https://github.com/homieiot/convention

Even if it is WIP, I have updated all the example files to the new API and in my preliminary test everything works fine :+1: But I would like to know for opinion about the new class names and functions. Don't hesitate to suggest or criticize whatever you want :stuck_out_tongue_closed_eyes: 

Best regards,

Aitor

bodiroga committed:
Move 'property' to its own file

bodiroga committed:
Move 'examples' to their own folder

bodiroga committed:
Add a method to check if datatype is valid

bodiroga committed:
Add more attributes to HomieNodeProperty

bodiroga committed:
Update HomieNodePropertyRange class

bodiroga committed:
Rename 'announce' methods to 'addProperty'

bodiroga committed:
Return 'property' object in the addProperty method

bodiroga committed:
Rename 'setProperty' to 'getProperty'

bodiroga committed:
Publish property attributes when nodes are published

bodiroga committed:
Rename 'sendProperties' to 'publishProperties'

bodiroga committed:
Properly implement fields getters and setters

bodiroga committed:
Add validation functions for format and units

bodiroga committed:
Fix error in isValidId function

bodiroga committed:
Improve format validation

bodiroga committed:
Check if datatype, format and unit are valid

bodiroga committed:
Fix property creation in node class

bodiroga committed:
Check if nodeId is valid

bodiroga committed:
Raise ValueError exception if property id is invalid

bodiroga committed:
Update the subscribe mechanism for properties

bodiroga committed:
Rename 'send' method to 'update'

bodiroga committed:
Rework main file

- Rename Homie class to Device.
- Update Node method to addNode.
- Update property subscription mechanism. callback(property, value).

bodiroga committed:
Publish 'stats' information

bodiroga committed:
Implement 'state' attributes

bodiroga committed:
Implement 'name' attribute for nodes

bodiroga committed:
Update homie version to 3.0.0

bodiroga committed:
Update all example files

jalmeroth commented:
Hi Aitor,

wow impressive work, once again! Have you seen my efforts in [dev](https://github.com/jalmeroth/homie-python/tree/dev) from 8 months ago? It might makes more sense to merge your changes here.

Let me know what you think.

Best regards,
Jan

bodiroga commented:
Hi Jan (@jalmeroth)!

Sorry, but I'm not sure if I have understood what you mean. Do you want me to make my changes against your dev branch? I don't know if that's going to be easy 😞 Anyway, we should talk about how to handle this situation: what Homie version should we support in the library? Homie v2? Homie v3? Both at the same time? What's the best way to deal it? Two repositories and two libraries (homie-python and homie3-python)? Everything in the same repository? What about the "pip" installation process? How can that be managed?

I think that I will wait until we decide something about this before changing anything in the PR 😉 

Best regards!

Aitor

bodiroga committed:
Update to Homie 3.0.1 and add more examples

bodiroga commented:
Hi @jalmeroth!

I have made a new commit to update the library to Homie 3.0.1 (they have introduced a new "$retained" attribute to the properties to make them stateless) and have added two new examples.

Let me know how do you want to address the different versions of the library (Homie 2 and Homie 3) and looking forward to seeing the library marked as stable :wink: 

Best regards

Aitor

bodiroga committed:
Fix $stats mismatch with specification

esdeboer commented:
Hi @bodiroga,

Shouldn't the isValidFormat check in helper.py check for none for everything except enum and color, as the format is not required for the other datatypes?  (or is it because of my old python 3.4 on my pi that I get a WARNING:homie.property:'None' is not a valid format for float)

Regards,
Eric

stefanroellin commented:
I have tested the branch and have found the following issues:

1. in the 3.0 specification "$online" is not used anymore, however there are still updates to this topic. Should probably be replaced with "$state".
2. during startup there is a warning that the deviceId is invalid, (since it is set to "None" in \_\_init\_\_)
3. personally I am not interested in the "signal" topic. Hence a possibility to exclude it would be good.

Thanks for your work!

rafdaems commented:
Thank you very much for this fork.
My openhab installations seems to autodetect mqtt subscribers created by your framework.

One question:
I don't understand the implementation of "settable".
see: https://homieiot.github.io/specification/ 
under 6.4.1 properties.

pls advise ?

ababilone commented:
Hi there,

Any news on this subject ? I'm also working on an OpenHAB 2.4 integration and this 3.0 binding is just what It needs. The work done by @bodiroga looks awesome.

Do you plan to merge it someday ?

Thanks for any feedback!

bodiroga committed:
Remove '' topic usage

Signed-off-by: Aitor Iturrioz <riturrioz@gmail.com>

bodiroga committed:
Remove "self.deviceID = None" statement

bodiroga committed:
Allow no format for certain datatypes

bodiroga commented:
Hi guys!

> Shouldn't the isValidFormat check in helper.py check for none for everything except enum and color, as the format is not required for the other datatypes?

Fixed @esdeboer!

> 1. in the 3.0 specification "$online" is not used anymore, however there are still updates to this topic. Should probably be replaced with "$state".

Fixed @stefanroellin!

> 2. during startup there is a warning that the deviceId is invalid, (since it is set to "None" in __init__)

Fixed!

> 3. personally I am not interested in the "signal" topic. Hence a possibility to exclude it would be good.

Yeah, you are right, but as Homie 4 is going to completely remove the "stats" information, I'm going to keep it for now. It doesn't cause any harm and someone can find it useful :wink: 

> pls advise ?

Hi @rafdaems! Could you please be more specific? I need to know first what is what you don't understand? :+1: 

Please, test the new changes and report if you find any error! Thanks for you help! :+1: 

Best regards,

Aitor

esdeboer commented:
Thanks for the fixes Aitor, looks good.
I don't get the deviceId warning anymore and I can add properties with a datatype without format.

Kixunil commented:
Hi, thanks a lot for this code! I used it for [`solight-homie`](https://github.com/ParalelnaPolis/solight-homie) and it works with OpenHAB. It's very nice how easily I could create a plugin for OpenHAB using this code.

stedon81 commented:
Hi,
I experienced some hang up issues with this homie implementation in this branch. Is it possible, that this library is not thread-safe? From my code, it looks like different threads would block each other if they try to do update or read operations at the same time. 
Regards,
stedon81

stedon81 commented:
OK, I found the answer - it's not homie, but the underlying paho mqtt library which is not thread safe:
https://github.com/eclipse/paho.mqtt.java/issues/453#issuecomment-402092568
Nevermind, sorry to have bothered :-)

mjcumming commented:
I have put together a python library for Homie 3 that I would like some feedback on. You can get it here

https://pypi.org/project/Homie3/

Mike

yozik04 commented:
I think this pull request needs to land to a separate repo and produce a separate pip package.
